luzer/luzer_args.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * SPDX-License-Identifier: ISC
+ *
+ * Copyright 2022-2023, Sergey Bronnikov
+ */
+
+#include <lauxlib.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "luzer_args.h"
+#include "macros.h"
+
+#define ENV_NOT_USE_CLI_ARGS "LUZER_NOT_USE_CLI_ARGS"
+
+#define SEED_CORPUS_PATH_FLAG "-corpus"
+
+#define FLAG_PATTERN_KEY "-%s="
+#define FLAG_PATTERN_KEY_VALUE "-%s=%s"
+#define FLAG_PATTERN_OVERHEAD 2
+#define FLAG_FULL_SIZE(flag) (strlen(flag) + FLAG_PATTERN_OVERHEAD)
+
+/* Structure for convenient argument parsing. */
+typedef struct {
+	char **argv;
+	int argc;
+} luzer_args;
+
+NO_SANITIZE static bool
+is_flag_in_args(luzer_args *f_args, const char *key) {
+	if (!f_args || !f_args->argv || f_args->argc <= 1) {
+                return false;
+        }
+
+        size_t flag_size = FLAG_FULL_SIZE(key);
+	char flag[flag_size + 1 /* null terminated */];
+	snprintf(flag, flag_size + 1, FLAG_PATTERN_KEY, key);
+	for (int i = 0; i < f_args->argc; i++) {
+		if (strncmp(f_args->argv[i], flag, flag_size) == 0) {
+                        return true;
+                }
+	}
+	return false;
+}
+
+NO_SANITIZE static int
+luaL_get_args_from_cli(lua_State *L, luzer_args *cli_args) {
+	lua_getglobal(L, "arg");
+
+	cli_args->argv = malloc(1 * sizeof(char*));
+	if (!cli_args->argv) {
+                return -1;
+        }
+
+	lua_pushnil(L);
+
+        bool use_cli_args = !getenv(ENV_NOT_USE_CLI_ARGS);
+	/* Zero arg is reserved for program name. */
+        cli_args->argc = 1;
+	while (lua_next(L, -2) != 0) {
+		const char *value = lua_tostring(L, -1);
+		const int key = lua_tointeger(L, -2);
+		lua_pop(L, 1);
+
+                if (key < 0) {
+                        continue;
+                }
+
+                const char *arg = strdup(value);
+                if (!arg) {
+                        return -1;
+                }
+
+                if (key == 0) {
+                        cli_args->argv[0] = (char*)arg;
+                        continue;
+                }
+
+                if (use_cli_args) {
+                        char **argvp = realloc(cli_args->argv, sizeof(char*) * (cli_args->argc + 1));
+                        if (argvp == NULL) {
+                                return -1;
+                        }
+                        cli_args->argv = argvp;
+                        cli_args->argv[cli_args->argc] = (char*)arg;
+                        cli_args->argc++;
+                }
+	}
+	lua_pop(L, 1);
+	return 0;
+}
+
+NO_SANITIZE static int
+luaL_get_args_from_table(lua_State *L, luzer_args *table_args, luzer_args *cli_args) {
+        if (lua_istable(L, -1) == 0) {
+                return -2;
+        }
+
+	lua_pushnil(L);
+
+	/* Processing a table with options. */
+	table_args->argc = 0;
+        while (lua_next(L, -2) != 0) {
+		const char *key = lua_tostring(L, -2);
+		const char *value = lua_tostring(L, -1);
+		lua_pop(L, 1);
+		if (is_flag_in_args(cli_args, key)) {
```

Comment:
this code should be moved to `merge_args`, isn't it?

Reply:
Yes, I corrected it